test/schedulers.py

```diff
@@ -0,0 +1,115 @@
+        the learning rate of the previous step, don't bother updating the
+        optimizer's learning rate.
+    summary_writer: Instance of `torch.utils.tensorboard.SummaryWriter`. If
+        provided, learning rate will be logged during calls to step if step
+        is called with write_to_summary=True. If summary_writer is None, then
+        no logging happens.
+  """
+  def __init__(self, optimizer, num_steps_per_epoch, divide_every_n_epochs=20,
+               divisor=5, num_warmup_epochs=1, min_delta_to_update_lr=0.000001,
+               summary_writer=None):
+    self._num_steps_per_epoch = num_steps_per_epoch
+    self._divide_every_n_epochs = divide_every_n_epochs
+    self._divisor = divisor
+    self._num_warmup_epochs = num_warmup_epochs
+    self._min_delta_to_update_lr = min_delta_to_update_lr
+    self._previous_lr = -1
+    self._max_lr = optimizer.param_groups[0]['lr']
+    self._summary_writer = summary_writer
+    super(WarmupAndExponentialDecayScheduler, self).__init__(optimizer)
+
+  def _epoch(self):
+    return self._step_count // self._num_steps_per_epoch
+
+  def get_lr(self):
+    epoch = self._epoch()
+    lr = 0.0
+
+    if epoch < self._num_warmup_epochs:
+      # Warmup epoch. Ramp up learning rate from 0.0 to self._max_lr using
+      # a linear slope.
+      num_warmup_steps = self._num_warmup_epochs * self._num_steps_per_epoch
+      lr = min(self._max_lr,
+               self._max_lr * ((self._step_count + 1.0) / num_warmup_steps))
+    else:
+      # Normal epoch. Use an exponential decay determined by init params.
+      lr = self._max_lr / (
+          self._divisor ** (epoch // self._divide_every_n_epochs))
+
+    # _LRScheduler expects a list of learning rates like this.
+    return [lr for _ in self.base_lrs]
+
+  def step(self, epoch=None):
+    epoch = self._epoch()
+    current_lr = self.get_lr()[0]
+
+    # Add current learning rate to Tensorboard metrics. For warmup epochs,
+    # log the learning rate at every step. For non-warmup epochs, log only
+    # the first step since the entire epoch will use the same learning rate.
+    if self._summary_writer:
+      if self._epoch() < self._num_warmup_epochs or (
+          self._step_count % self._num_steps_per_epoch == 0):
+        test_utils.add_scalar_to_summary(self._summary_writer,
+                                         'LearningRate',
+                                         current_lr,
+                                         self._step_count)
+
+    # Outside of warmup epochs, we use the same learning rate for every step
+    # in an epoch. Don't bother updating learning rate if it hasn't changed. 
+    if abs(current_lr - self._previous_lr) < self._min_delta_to_update_lr:
+      self._step_count += 1  # This normally happens in super().step().
+      return
+    self._previous_lr = current_lr
+    super(WarmupAndExponentialDecayScheduler, self).step()
+
+
+def wrap_optimizer_with_scheduler(optimizer, flags, num_steps_per_epoch=None,
```

Comment:
I think it is better, instead of passing `FLAGS`, to pass the single configurables directly in.

Reply:
Agreed. I was trying to minimize lines in test_train_imagenet but passing FLAGS is not a good idea. Switched to individual args